src/metaopt/io/database/mongodb.py

```diff
@@ -0,0 +1,277 @@
+            is_locked = self._conn.is_locked
+        except pymongo.errors.ConnectionFailure as e:
+            self._logger.debug("Could not determine if connection is locked.")
+            raise_from(DatabaseError("Could not determine if connection is locked."), e)
+        return is_locked
+
+    def lock(self):
+        """Lock a database."""
+        try:
+            lock = self._conn.fsync(lock=True)
+        except pymongo.errors.ConnectionFailure as e:
+            self._logger.debug("Could not lock a database")
+            raise_from(DatabaseError("Could not determine lock due to connection."), e)
+        return lock
+
+    def unlock(self):
+        """Unlock a database."""
+        try:
+            unlock = self._conn.unlock()
+        except pymongo.errors.ConnectionFailure as e:
+            self._logger.debug("Could not determine unlock due to connection")
+            raise_from(DatabaseError("Could not determine unlock due to connection."), e)
+        return unlock
+
+    def write(self, collection_name, data,
+              query=None, atomic=False, selection=None):
+        """Write new information to a collection. Perform insert or update.
+
+        Arguments
+        ---------
+        collection_name : str
+           A collection inside database, a table.
+        data : dict or list of dicts
+           New data that will **be inserted** or that will **update** entries.
+        query : dict, optional
+           Assumes an update operation: filter entries in collection to be updated.
+        atomic : bool, optional
+           If True, update only the most recent entry that matches `query`.
+        selection : dict, optional
+           Used in `atomic` update operation: elements of the matched entry to
+           return, the projection.
+
+        :return: operation success, if not `atomic`, matched updated document else.
+
+        .. note::
+           In the case of an insert operation, `data` variable will be updated
+           to contain a unique *_id* key.
+
+        .. note::
+           In the case of an update operation, if `query` fails to find a
+           document that matches, insert of `data` will be performed instead.
+        """
+        dbcollection = self._db[collection_name]
+
+        if query is None:
+            # We can assume that we do not want to update.
+            # So we do insert_many instead.
+            if type(data) not in (list, tuple):
+                data = [data]
+            result = dbcollection.insert_many(documents=data)
+            return result.acknowledged
+
+        update_data = {'$set': data}
+        if atomic is False:
+            result = dbcollection.update_many(filter=query,
+                                              update=update_data,
+                                              upsert=True)
+            return result.acknowledged
+        elif atomic is True:
+            selected_one_key, selection = _get_selected_one_key(selection)
```

Comment:
Ok I understand the intent. Then we would not need to do `db.read({}, {'arg':1})['arg']`. But isn't a bit weird? I'd be surprised as a user if making a selection of a single value suddenly returns items rather than dictionaries. It's convenient, but it makes the method's behavior more complex. What if some function makes dynamic reads and at some point its dictionary gets unitary. It would need to take into account that at if its dictionary gets unitary than it needs to rebuild the dictionary because the method suddenly returns an item.

Personally, I would favor coherence and simplicity over a convenience like this. Also, accumulation of small tweaks like that adds on the burden of testing and maintenance, thus slowing down development. Writing `db.read({}, {'arg':1})['arg']` is less elegant but doesn't have such an impact.

Reply:
With more dotted selections 'arg1.arg2.arg3' the more ugly it will get. But I agree about your point on coherence and simplicity. Will see what I can do 😉

Reply:
We should finish this [discussion](https://github.com/mila-udem/metaopt/pull/8#discussion_r153380279) before spending time on `selected_one_key`, otherwise it might be a waste of time.